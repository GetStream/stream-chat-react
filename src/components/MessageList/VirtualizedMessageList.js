// @ts-check
import React, {
  useContext,
  useState,
  useEffect,
  useCallback,
  useRef,
} from 'react';
import { Virtuoso } from 'react-virtuoso';

import { smartRender } from '../../utils';
import MessageNotification from './MessageNotification';
import { ChannelContext, TranslationContext } from '../../context';
import { EventComponent } from '../EventComponent';
import { LoadingIndicator as DefaultLoadingIndicator } from '../Loading';
import { EmptyStateIndicator as DefaultEmptyStateIndicator } from '../EmptyStateIndicator';
import { TypingIndicator as DefaultTypingIndicator } from '../TypingIndicator';
import {
  FixedHeightMessage,
  MessageDeleted as DefaultMessageDeleted,
} from '../Message';

/**
 * VirtualizedMessageList - This component renders a list of messages in a virtual list. Its a consumer of [Channel Context](https://getstream.github.io/stream-chat-react/#channel)
 * It is pretty fast for rendering thousands of messages but it needs its Message componet to have fixed height
 * @example ../../docs/VirtualizedMessageList.md
 * @type {React.FC<import('types').VirtualizedMessageListInternalProps>}
 */
const VirtualizedMessageList = ({
  client,
  messages,
  loadMore,
  hasMore,
  loadingMore,
  messageLimit = 100,
  overscan = 200,
  scrollSeekPlaceHolder,
  Message = FixedHeightMessage,
  MessageSystem = EventComponent,
  MessageDeleted = DefaultMessageDeleted,
  TypingIndicator = DefaultTypingIndicator,
  LoadingIndicator = DefaultLoadingIndicator,
  EmptyStateIndicator = DefaultEmptyStateIndicator,
}) => {
  const { t } = useContext(TranslationContext);
  const [newMessagesNotification, setNewMessagesNotification] = useState(false);

  const virtuoso = useRef(
    /** @type {import('react-virtuoso').VirtuosoMethods | undefined} */ (undefined),
  );
  const mounted = useRef(false);
  const atBottom = useRef(false);
  const lastMessageId = useRef('');

  useEffect(() => {
    /* handle scrolling behavior for new messages */
    if (!messages.length) return;

    const lastMessage = messages[messages.length - 1];
    const prevMessageId = lastMessageId.current;
    lastMessageId.current = lastMessage.id || ''; // update last message id

    /* do nothing if new messages are loaded from top(loadMore)  */
    if (lastMessage.id === prevMessageId) return;

    /* if list is already at the bottom make it sticky */
    if (atBottom.current) {
      setTimeout(() => virtuoso.current?.scrollToIndex(messages.length)); // setTimeout is needed to delay the scroll until react flushes
      return;
    }

    /* if the new message belongs to current user scroll to bottom */
    if (lastMessage.user?.id === client.userID) {
      setTimeout(() => virtuoso.current?.scrollToIndex(messages.length));
      return;
    }

    /* otherwise just show newMessage notification  */
    setNewMessagesNotification(true);
  }, [client.userID, messages]);

  useEffect(() => {
    /*
     * scroll to bottom when list is rendered for the first time
     * this is due to initialTopMostItemIndex buggy behavior leading to empty screen
     */
    if (mounted.current) return;
    mounted.current = true;
    if (messages.length && virtuoso.current) {
      virtuoso.current.scrollToIndex(messages.length - 1);
    }
  }, [messages.length]);

  const messageRenderer = useCallback(
    (message) => {
      if (!message) return <></>;

      if (message.type === 'channel.event' || message.type === 'system')
        return <MessageSystem message={message} />;

      if (message.deleted_at)
        return smartRender(MessageDeleted, { message }, null);

      return <Message message={message} />;
    },
    [MessageDeleted],
  );

  return (
    <div className="str-chat__virtual-list">
      <Virtuoso
        // @ts-ignore
        ref={virtuoso}
        totalCount={messages.length}
        overscan={overscan}
        scrollSeek={scrollSeekPlaceHolder}
        item={(i) => messageRenderer(messages[i])}
        emptyComponent={() => <EmptyStateIndicator listType="message" />}
        header={() => (
          <div
            className="str-chat__virtual-list__loading"
            style={{ visibility: loadingMore ? undefined : 'hidden' }}
          >
            <LoadingIndicator size={20} />
          </div>
        )}
        footer={() => TypingIndicator && <TypingIndicator avatarSize={24} />}
        startReached={() => {
          // mounted.current prevents immediate loadMore on first render
          if (mounted.current && hasMore) {
            loadMore(messageLimit).then(
              virtuoso.current?.adjustForPrependedItems,
            );
          }
        }}
        atBottomStateChange={(isAtBottom) => {
          atBottom.current = isAtBottom;
          if (isAtBottom && newMessagesNotification)
            setNewMessagesNotification(false);
        }}
      />

      <div className="str-chat__list-notifications">
        <MessageNotification
          showNotification={newMessagesNotification}
          onClick={() => {
            if (virtuoso.current)
              virtuoso.current.scrollToIndex(messages.length);
            setNewMessagesNotification(false);
          }}
        >
          {t('New Messages!')}
        </MessageNotification>
      </div>
    </div>
  );
};

// TODO: fix the types here when everything converted to proper TS
/**
 * @param {import("types").VirtualizedMessageListProps} props
 * @returns {React.ElementType<import("types").VirtualizedMessageListInternalProps>}
 */
export default function VirtualizedMessageListWithContext(props) {
  // @ts-ignore
  return (
    <ChannelContext.Consumer>
      {(
        /* {Required<Pick<import('types').ChannelContextValue, 'client' | 'messages' | 'loadMore' | 'hasMore' | 'loadingMore'>>} */ context,
      ) => (
        <VirtualizedMessageList
          client={context.client}
          // @ts-ignore
          messages={context.messages}
          // @ts-ignore
          loadMore={context.loadMore}
          // @ts-ignore
          hasMore={context.hasMore}
          // @ts-ignore
          loadingMore={context.loadingMore}
          {...props}
        />
      )}
    </ChannelContext.Consumer>
  );
}
