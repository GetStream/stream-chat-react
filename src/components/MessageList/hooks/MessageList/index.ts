export * from './useEnrichedMessages';
export * from './useMessageListElements';
export * from './useMessageListScrollManager';
export * from './useScrollLocationLogic';
