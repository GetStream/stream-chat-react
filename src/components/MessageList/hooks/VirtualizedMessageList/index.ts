export * from './useGiphyPreview';
export * from './useMessageSetKey';
export * from './useNewMessageNotification';
export * from './usePrependMessagesCount';
export * from './useScrollToBottomOnNewMessage';
export * from './useShouldForceScrollToBottom';
