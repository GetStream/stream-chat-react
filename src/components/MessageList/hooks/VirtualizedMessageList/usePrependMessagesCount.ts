import { useMemo, useRef } from 'react';

import type { StreamMessage } from '../../../../context/ChannelStateContext';

import type { DefaultStreamChatGenerics } from '../../../../types/types';

const STATUSES_EXCLUDED_FROM_PREPEND = ({
  failed: true,
  sending: true,
} as const) as Record<string, boolean>;

export function usePrependedMessagesCount<
  StreamChatGenerics extends DefaultStreamChatGenerics = DefaultStreamChatGenerics
>(messages: StreamMessage<StreamChatGenerics>[], hasDateSeparator: boolean) {
  const firstRealMessageIndex = hasDateSeparator ? 1 : 0;
  const firstMessageOnFirstLoadedPage = useRef<StreamMessage<StreamChatGenerics>>();
  const previousFirstMessageOnFirstLoadedPage = useRef<StreamMessage<StreamChatGenerics>>();
  const previousNumItemsPrepended = useRef(0);

  const numItemsPrepended = useMemo(() => {
    if (!messages || !messages.length) {
      previousNumItemsPrepended.current = 0;
      return 0;
    }

    const currentFirstMessage = messages?.[firstRealMessageIndex];

    const noNewMessages =
      currentFirstMessage?.id === previousFirstMessageOnFirstLoadedPage.current?.id;

    // This is possible only, when sending messages very quickly (basically single char messages submitted like a crazy) in empty channel (first page)
    // Optimistic UI update, when sending messages, can lead to a situation, when
    // the order of the messages changes for a moment. This can happen, when a user
    // sends multiple messages withing few milliseconds. E.g. we send a message A
    // then message B. At first we have message array with both messages of status "sending"
    // then response for message A is received with a new - later - created_at timestamp
    // this leads to rearrangement of 1.B ("sending"), 2.A ("received"). Still firstMessageOnFirstLoadedPage.current
    // points to message A, but now this message has index 1 => previousNumItemsPrepended.current === 1
    // That in turn leads to incorrect index calculation in VirtualizedMessageList trying to access a message
    // at non-existent index. Therefore, we ignore messages of status "sending" / "failed" in order they are
    // not considered as prepended messages.
    const firstMsgMovedAfterMessagesInExcludedStatus = !!(
      currentFirstMessage?.status && STATUSES_EXCLUDED_FROM_PREPEND[currentFirstMessage.status]
    );

    if (noNewMessages || firstMsgMovedAfterMessagesInExcludedStatus) {
      return previousNumItemsPrepended.current;
    }

    if (!firstMessageOnFirstLoadedPage.current) {
      firstMessageOnFirstLoadedPage.current = currentFirstMessage;
    }
    previousFirstMessageOnFirstLoadedPage.current = currentFirstMessage;
    // if new messages were prepended, find out how many
    // start with this number because there cannot be fewer prepended items than before
    for (
      let prependedMessageCount = previousNumItemsPrepended.current;
      prependedMessageCount < messages.length;
      prependedMessageCount += 1
    ) {
      const messageIsFirstOnFirstLoadedPage =
        messages[prependedMessageCount].id === firstMessageOnFirstLoadedPage.current?.id;

      if (messageIsFirstOnFirstLoadedPage) {
        previousNumItemsPrepended.current = prependedMessageCount - firstRealMessageIndex;
        return previousNumItemsPrepended.current;
      }
    }

    // if no match has found, we have jumped - reset the prepended item count.
    firstMessageOnFirstLoadedPage.current = currentFirstMessage;
    previousNumItemsPrepended.current = 0;
    return 0;
    // TODO: there's a bug here, the messages prop is the same array instance (something mutates it)
    // that's why the second dependency is necessary
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, [firstRealMessageIndex, messages, messages?.length]);

  return numItemsPrepended;
}
