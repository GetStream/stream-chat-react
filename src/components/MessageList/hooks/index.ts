export * from './useEnrichedMessages';
export * from './useLastReadData';
export * from './useMessageListElements';
export * from './useMessageListScrollManager';
export * from './useNewMessageNotification';
export * from './usePrependMessagesCount';
export * from './useScrollLocationLogic';
export * from './useShouldForceScrollToBottom';
