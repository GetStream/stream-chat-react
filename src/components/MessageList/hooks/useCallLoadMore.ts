import { useCallback } from 'react';

import type { MessageListProps } from '../MessageList';

export const useCallLoadMore = (loadMore: MessageListProps['loadMore'], limit: number) =>
  useCallback(() => {
    if (loadMore) {
      loadMore(limit);
    }
  }, [limit, loadMore]);
