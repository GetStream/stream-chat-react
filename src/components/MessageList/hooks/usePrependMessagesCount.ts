import { useMemo, useRef } from 'react';

import type { StreamMessage } from '../../../context/ChannelStateContext';

import type { DefaultStreamChatGenerics } from '../../../types/types';

const STATUSES_EXCLUDED_FROM_PREPEND = ['sending', 'failed'];

export function usePrependedMessagesCount<
  StreamChatGenerics extends DefaultStreamChatGenerics = DefaultStreamChatGenerics
>(messages: StreamMessage<StreamChatGenerics>[], hasDateSeparator: boolean) {
  const firstRealMessageIndex = hasDateSeparator ? 1 : 0;
  const firstMessageId = useRef<string>();
  const earliestMessageId = useRef<string>();
  const previousNumItemsPrepended = useRef(0);

  const numItemsPrepended = useMemo(() => {
    if (!messages || !messages.length) {
      previousNumItemsPrepended.current = 0;
      return 0;
    }

    const currentFirstMessageId = messages?.[firstRealMessageIndex]?.id;
    // if no new messages were prepended, return early (same amount as before)
    if (currentFirstMessageId === earliestMessageId.current) {
      return previousNumItemsPrepended.current;
    }

    if (!firstMessageId.current) {
      firstMessageId.current = currentFirstMessageId;
    }
    earliestMessageId.current = currentFirstMessageId;
    // if new messages were prepended, find out how many
    // start with this number because there cannot be fewer prepended items than before
    let adjustPrependedMessageCount = 0;
    for (let i = previousNumItemsPrepended.current; i < messages.length; i += 1) {
      // Optimistic UI update, when sending messages, can lead to a situation, when
      // the order of the messages changes for a moment. This can happen, when a user
      // sends multiple messages withing few milliseconds. E.g. we send a message A
      // then message B. At first we have message array with both messages of status "sending"
      // then response for message A is received with a new - later - created_at timestamp
      // this leads to rearrangement of 1.B ("sending"), 2.A ("received"). Still firstMessageId.current
      // points to message A, but now this message has index 1 => previousNumItemsPrepended.current === 1
      // That in turn leads to incorrect index calculation in VirtualizedMessageList trying to access a message
      // at non-existent index. Therefore, we ignore messages of status "sending" / "failed" in order they are
      // not considered as prepended messages.
      if (
        messages[i]?.status &&
        // eslint-disable-next-line @typescript-eslint/no-non-null-assertion
        STATUSES_EXCLUDED_FROM_PREPEND.includes(messages[i].status!) &&
        messages[i].id !== firstMessageId.current
      ) {
        adjustPrependedMessageCount++;
      }
      if (messages[i].id === firstMessageId.current) {
        previousNumItemsPrepended.current = i - adjustPrependedMessageCount;
        return previousNumItemsPrepended.current;
      }
    }

    // if no match has found, we have jumped - reset the prepended item count.
    firstMessageId.current = currentFirstMessageId;
    previousNumItemsPrepended.current = 0;
    return 0;
    // TODO: there's a bug here, the messages prop is the same array instance (something mutates it)
    // that's why the second dependency is necessary
  }, [messages, messages?.length]);

  return numItemsPrepended;
}
