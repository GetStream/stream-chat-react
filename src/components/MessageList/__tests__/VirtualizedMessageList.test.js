import React from 'react';
import { cleanup } from '@testing-library/react';
import renderer from 'react-test-renderer';
import '@testing-library/jest-dom';

import {
  generateChannel,
  generateMember,
  generateMessage,
  generateUser,
  getOrCreateChannelApi,
  getTestClientWithUser,
  useMockedApis,
} from '../../../mock-builders';

import { usePrependedMessagesCount } from '../hooks/usePrependMessagesCount';
import { VirtualizedMessageList } from '../VirtualizedMessageList';

import { Chat } from '../../Chat';
import { Channel } from '../../Channel';

jest.mock('react-virtuoso', () => {
  const { Virtuoso } = jest.requireActual('react-virtuoso');
  const { forwardRef } = jest.requireActual('react');
  return {
    Virtuoso: forwardRef((props, ref) => (
      <Virtuoso
        ref={ref}
        {...props}
        fixedItemHeight={30}
        initialTopMostItemIndex={0}
        overscan={0}
      />
    )),
  };
});

jest.mock('../../Loading', () => ({
  LoadingIndicator: jest.fn(() => <div>LoadingIndicator</div>),
}));

jest.mock('../../Message', () => ({
  FixedHeightMessage: jest.fn(({ groupedByUser }) => (
    <div data-testid='msg'>
      FixedHeightMessage groupedByUser:
      {groupedByUser ? 'true' : 'false'}
    </div>
  )),
}));

async function createChannel(empty = false) {
  const user1 = generateUser();
  const user2 = generateUser();
  const mockedChannel = generateChannel({
    members: [generateMember({ user: user1 }), generateMember({ user: user2 })],
    messages: empty
      ? []
      : ' '
          .repeat(20)
          .split(' ')
          .map((v, i) => generateMessage({ user: i % 3 ? user1 : user2 })),
  });
  const client = await getTestClientWithUser({ id: 'id' });
  useMockedApis(client, [getOrCreateChannelApi(mockedChannel)]); // eslint-disable-line react-hooks/rules-of-hooks
  const channel = client.channel('messaging', mockedChannel.id);
  await channel.watch();

  return { channel, client };
}

// simple test since Virtuoso heavily relies on document height and jsdom doesn't support it
describe('VirtualizedMessageList', () => {
  afterEach(cleanup);
  beforeEach(jest.clearAllMocks);

  it('should render the list without any message', async () => {
    const { channel, client } = await createChannel(true);
    let tree;

    function createNodeMock(element) {
      if (element.type === 'div') {
        return {
          addEventListener() {},
          removeEventListener() {},
        };
      }
      return null;
    }

    await renderer.act(() => {
      tree = renderer.create(
        <Chat client={client}>
          <Channel channel={channel}>
            <VirtualizedMessageList />
          </Channel>
        </Chat>,
        {
          createNodeMock,
        },
      );
    });

    expect(tree.toJSON()).toMatchSnapshot();
  });
});

describe('usePrependedMessagesCount', () => {
  const TestCase = ({ messages }) => {
    const prependCount = usePrependedMessagesCount(messages);
    return <div>{prependCount}</div>;
  };

  const expectPrependCount = (count, root) => {
    expect(root.findByType('div').props.children).toStrictEqual(count);
  };

  it('determines 0 prepended messages for empty message list', async () => {
    const render = await renderer.create(<TestCase messages={[]} />);

    expectPrependCount(0, render.root);
  });

  const messageBatch = (ids, status) => ids.map((id) => ({ id, status }));
  const firstBatch = (status) => messageBatch(['a'], status);
  const secondBatch = (status) => messageBatch(['c', 'b'], status);
  const thirdBatch = (status) => messageBatch(['e', 'd'], status);

  const testPrependCount = async (status, first, second, third) => {
    const firstMessage = firstBatch(status);
    const secondMsgBatch = secondBatch(status);
    const thirdMsgBatch = thirdBatch(status);

    const render = await renderer.create(<TestCase messages={[]} />);

    await renderer.act(async () => {
      await render.update(<TestCase messages={[...firstMessage]} />);
      expectPrependCount(first, render.root);
    });

    await renderer.act(async () => {
      await render.update(<TestCase messages={[...secondMsgBatch, ...firstMessage]} />);
      expectPrependCount(second, render.root);
    });

    await renderer.act(async () => {
      await render.update(
        <TestCase messages={[...thirdMsgBatch, ...secondMsgBatch, ...firstMessage]} />,
      );
      expectPrependCount(third, render.root);
    });
  };

  it('calculates the prepended count for messages of status "received"', async () => {
    await testPrependCount('received', 0, 2, 4);
  });

  it('ignores the messages of status "sending" from the prepended messages count', async () => {
    await testPrependCount('sending', 0, 0, 0);
  });

  it('ignores the messages of status "failed" from the prepended messages count', async () => {
    await testPrependCount('failed', 0, 0, 0);
  });

  it('calculates the prepended messages count for the messages of status undefined', async () => {
    await testPrependCount(undefined, 0, 2, 4);
  });
});
