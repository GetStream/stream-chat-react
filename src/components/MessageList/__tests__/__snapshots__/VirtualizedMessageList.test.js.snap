// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`VirtualizedMessageList should render the list without any message 1`] = `
<div
  className="str-chat messaging light str-chat-channel str-chat__channel"
>
  <div
    className="str-chat__container"
  >
    <div
      className="str-chat__virtual-list"
    >
      <div
        className="str-chat__message-list-scroll"
        data-test-id="virtuoso-scroller"
        data-virtuoso-scroller={true}
        style={
          Object {
            "WebkitOverflowScrolling": "touch",
            "height": "100%",
            "outline": "none",
            "overflowX": "hidden",
            "overflowY": "auto",
            "position": "relative",
          }
        }
        tabIndex={0}
      >
        <div
          data-viewport-type="element"
          style={
            Object {
              "height": "100%",
              "position": "absolute",
              "top": 0,
              "width": "100%",
            }
          }
        >
          <div />
          <div
            className="str-chat__empty-channel"
          >
            <svg
              data-testid="chat-bubble"
              fill="none"
              height="96"
              viewBox="0 0 136 136"
              width="96"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M106 24.5H30C24.775 24.5 20.5 28.775 20.5 34V119.5L39.5 100.5H106C111.225 100.5 115.5 96.225 115.5 91V34C115.5 28.775 111.225 24.5 106 24.5ZM106 91H39.5L30 100.5V34H106V91Z"
                fill="#B4B7BB"
              />
            </svg>
            <p
              className="str-chat__empty-channel-text"
              role="listitem"
            >
              No chats here yetâ€¦
            </p>
          </div>
          <div />
        </div>
      </div>
    </div>
    <div
      className="str-chat__list-notifications"
    />
  </div>
</div>
`;
